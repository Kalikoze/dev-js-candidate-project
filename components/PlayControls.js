import React, { Component } from 'react';

export default class PlayControls extends Component {
  constructor() {
    super();
    this.state = {
      track: 1,
      isPlaying: true
    };
  }

  componentDidMount() {
    this.audio.volume = 0.6;
  }

  render() {
    const { track, isPlaying } = this.state;

    return (
      <section className='play-controls'>
        <audio src={`../static/music/track${track}.mp3`} ref={node => this.audio = node} autoPlay loop />
        <img src='../static/play-controls/rewind.svg'
          onClick={() => {
            if(track === 1) {
              return this.setState({ track: 4, isPlaying: true });
            }
            this.setState({track: track - 1, isPlaying: true});
          }}
        />
        <img
          src={isPlaying ? '../static/play-controls/pause-button.svg' : '../static/play-controls/play-button.svg'}
          onClick={() => (isPlaying ? this.audio.pause() : this.audio.play(), this.setState({ isPlaying: !isPlaying }))}
        />
        <img src='../static/play-controls/fast-forward.svg'
          onClick={() => {
            if(track === 4) {
              return this.setState({ track: 1, isPlaying: true });
            }
            this.setState({track: track + 1, isPlaying: true});
          }}
        />
        <style jsx>{`
          .play-controls {
            position: absolute;
            right: 20px;
            top: 10px;
          }

          img {
            cursor: pointer;
            height: 40px;
            margin: 20px;
            opacity: 0.7;
            width: 40px;
          }

          img:hover {
            opacity: 1;
          }

          img:active {
            transform: scale(1.2);
          }
        `}</style>
      </section>
    );
  }
}
