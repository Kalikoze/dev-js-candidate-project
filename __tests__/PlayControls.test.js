import React from 'react';
import PlayControls from '../components/PlayControls';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { mount, configure } from 'enzyme';

configure({ adapter: new Adapter() });

describe('PlayControls component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(<PlayControls />);
  });

  it('should render the PlayControls correctly with the right classes and styles with a vaild ref', () => {
    const createNodeMock = (element) => {
      if (element.type === 'audio') {
        return {
          focus() {},
        };
      }
      return null;
    };

    const tree = renderer.create(wrapper, { createNodeMock }).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should be defined', () => {
    expect(wrapper).toBeDefined();
  });

  it('the state of isPlaying should toggle on click of Play/Pause', () => {
    expect(wrapper.state().isPlaying).toEqual(true);

    const button = wrapper.find('img').at(1);

    // Terminal error: Error: Not implemented: HTMLMediaElement.prototype.pause
    // Still passes test and changes state.
    button.simulate('click');
    expect(wrapper.state().isPlaying).toEqual(false);
    button.simulate('click');
    expect(wrapper.state().isPlaying).toEqual(true);
  });

  it('should change track number in state depending on click of Fast-Forward or Rewind', () => {
    expect(wrapper.state().track).toEqual(1);

    const rewind = wrapper.find('img').at(0);
    const fastForward = wrapper.find('img').at(2);

    rewind.simulate('click');
    expect(wrapper.state().track).toEqual(4);
    fastForward.simulate('click');
    fastForward.simulate('click');
    fastForward.simulate('click');
    expect(wrapper.state().track).toEqual(3);
  });
});
