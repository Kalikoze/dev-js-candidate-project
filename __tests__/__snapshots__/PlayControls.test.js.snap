// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayControls component should render the PlayControls correctly with the right classes and styles with a vaild ref 1`] = `
<section
  className="jsx-1701706752 play-controls"
>
  <audio
    autoPlay={true}
    className="jsx-1701706752"
    loop={true}
    src="../static/music/track1.mp3"
  />
  <img
    className="jsx-1701706752"
    onClick={[Function]}
    src="../static/play-controls/rewind.svg"
  />
  <img
    className="jsx-1701706752"
    onClick={[Function]}
    src="../static/play-controls/pause-button.svg"
  />
  <img
    className="jsx-1701706752"
    onClick={[Function]}
    src="../static/play-controls/fast-forward.svg"
  />
</section>
`;
